File: spring-webflux/src/main/java/victor/training/reactive/usecase/complex/ExternalAPIs.java
package victor.training.reactive.usecase.complex;

import lombok.extern.slf4j.Slf4j;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Slf4j
class ExternalAPIs {

   public static Mono<Void> auditResealedProduct(Product product) {
      // if (true) return Mono.error(new RuntimeException("Sh*t happens..."));
      return WebClient.create().get().uri("http://localhost:9999/api/audit-resealed/" + product)
              .retrieve()
              .toBodilessEntity()
              .doOnSubscribe(s -> log.info("Calling Audit REST: " +product.getId()))
              .then();
   }

   public static Mono<ProductRatingResponse> getProductRating(long productId) {
      // if (true) return Mono.error(new RuntimeException("Sh*t happens..."));

      return WebClient.create().get().uri("http://localhost:9999/api/rating/{}", productId)
              .retrieve()
              .bodyToMono(ProductRatingResponse.class)
              .doOnSubscribe(s -> log.info("Calling getProductRating API: " + productId))
              ;
   }


}

File: spring-webflux/src/main/java/victor/training/reactive/usecase/complex/ExternalAPIs.kt
package victor.training.reactive.usecase.complex

import org.reactivestreams.Subscription
import org.slf4j.LoggerFactory
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono

 object ExternalAPIs {
    private val log = LoggerFactory.getLogger(ExternalAPIs::class.java)
    fun auditResealedProduct(product: Product): Mono<Void> {
        // if (true) return Mono.error(new RuntimeException("Sh*t happens..."));
        return WebClient.create().get().uri("http://localhost:9999/api/audit-resealed/$product")
            .retrieve()
            .toBodilessEntity()
            .doOnSubscribe { s: Subscription? -> log.info("Calling Audit REST: " + product.id) }
            .then() //                .delaySubscription(Duration.ofMillis(Random.Default.nextLong(100)))
    }

    fun getProductRating(productId: Long): Mono<ProductRatingResponse> {
        return if (Math.random() < .5) {
            Mono.error(IllegalArgumentException("INTENTIONAT"))
        } else WebClient.create().get().uri("http://localhost:9999/api/rating/{}", productId)
            .retrieve()
            .bodyToMono(ProductRatingResponse::class.java)
            .doOnSubscribe { s: Subscription? -> log.info("BUN: Calling getProductRating API: $productId") }
        //        log.debug("PROST/MINCINOS: Now calling get product rating " + productId);
    }
}